internal/cadence/linter.go

```diff
@@ -0,0 +1,351 @@
+/*
+ * Flow CLI
+ *
+ * Copyright 2019 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cadence
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	cadenceLint "github.com/onflow/cadence-tools/lint"
+	cdcTests "github.com/onflow/cadence-tools/test/helpers"
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/errors"
+	"github.com/onflow/cadence/runtime/parser"
+	"github.com/onflow/cadence/runtime/sema"
+	"github.com/onflow/cadence/runtime/stdlib"
+	"github.com/onflow/cadence/tools/analysis"
+	"github.com/onflow/flowkit/v2"
+	"golang.org/x/exp/maps"
+)
+
+type linter struct {
+	checkers              map[string]*sema.Checker
+	state                 *flowkit.State
+	checkerStandardConfig *sema.Config
+	checkerScriptConfig   *sema.Config
+}
+
+type positionedError interface {
+	error
+	ast.HasPosition
+}
+
+// Error diagnostic categories
+const (
+	SemanticErrorCategory = "semantic-error"
+	SyntaxErrorCategory   = "syntax-error"
+	ErrorCategory         = "error"
+)
+
+var analyzers = maps.Values(cadenceLint.Analyzers)
+
+func newLinter(state *flowkit.State) *linter {
+	l := &linter{
+		checkers: make(map[string]*sema.Checker),
+		state:    state,
+	}
+
+	// Create checker configs for both standard and script
+	// Scripts have a different stdlib than contracts and transactions
+	l.checkerStandardConfig = l.newCheckerConfig(newStandardLibrary())
+	l.checkerScriptConfig = l.newCheckerConfig(newScriptStandardLibrary())
+
+	return l
+}
+
+func (l *linter) lintFile(
+	filePath string,
+) (
+	[]analysis.Diagnostic,
+	error,
+) {
+	diagnostics := make([]analysis.Diagnostic, 0)
+	location := common.StringLocation(filePath)
+
+	code, err := l.state.ReadFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+	codeStr := string(code)
+
+	program, parseErr := parser.ParseProgram(nil, code, parser.Config{})
+	if parseErr != nil {
+		errorDiagnostics, err := getDiagnosticsForParsingCheckingError(parseErr, location, codeStr)
+		if err != nil {
+			return nil, err
+		}
+
+		diagnostics = append(diagnostics, errorDiagnostics...)
+	}
+
+	// If the program is nil, nothing can be checked & analyzed so return early
+	if program == nil {
+		return diagnostics, nil
+	}
+
```

Comment:
lets make a comment that we are doing both a Parse and also a checker(?), and just briefly describe how both of them are part of what this "linter" function is